follower/source.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package follower
+
+import (
+	"fmt"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage/badger/operation"
+
+	"github.com/optakt/flow-dps/models/dps"
+)
+
+type executionFollower interface {
+	Update() (*ledger.TrieUpdate, error)
+	Header(height uint64) (*flow.Header, error)
+	Collections(height uint64) ([]*flow.LightCollection, error)
+	Guarantees(height uint64) ([]*flow.CollectionGuarantee, error)
+	Seals(height uint64) ([]*flow.Seal, error)
+	Transactions(height uint64) ([]*flow.TransactionBody, error)
+	Results(height uint64) ([]*flow.TransactionResult, error)
+	Events(height uint64) ([]flow.Event, error)
+}
+
+type consensusFollower interface {
+	Height() uint64
+	BlockID() flow.Identifier
+}
+
+type Source struct {
+	log zerolog.Logger
+
+	db        *badger.DB
+	execution executionFollower
+	consensus consensusFollower
+
+	blockID flow.Identifier
+	height  uint64
+}
+
+func NewSource(log zerolog.Logger, execution executionFollower, consensus consensusFollower, db *badger.DB) *Source {
+	s := Source{
+		log: log,
+
+		db:        db,
+		execution: execution,
+		consensus: consensus,
+
+		blockID: flow.ZeroID,
+	}
+
+	return &s
+}
+
+func (s *Source) Update() (*ledger.TrieUpdate, error) {
+	update, err := s.execution.Update()
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve trie update: %w", err)
+	}
```

Comment:
So here, I feel like the boundaries of the components are not well thought-out. The source component should, imo, be much smarter than this. It should orchestrate a dumb, stupid execution follower. It should load all trie updates from the execution follower, and then, if all of them have been consumed, it should tell the execution follower to switch to the next block.